pyccel/codegen/printing/ccode.py

```diff
@@ -586,6 +602,68 @@ def function_signature(self, expr):
         else:
             return '{0}{1}({2})'.format(ret_type, name, arg_code)
 
+    def _print_IndexedElement(self, expr):
+        if isinstance(expr.base, IndexedVariable):
+            base = expr.base.internal_variable
+        else:
+            base = expr.base
+        base = self._print(expr.base.label)
+        inds = list(expr.indices)
+        inds = inds[::-1]
+        base_shape = Shape(expr.base)
+        allow_negative_indexes = (isinstance(expr.base, IndexedVariable) and \
+                expr.base.internal_variable.allows_negative_indexes)
+        for i, ind in enumerate(inds):
+            if isinstance(ind, PyccelUnarySub) and isinstance(ind.args[0], Integer):
+                inds[i] = PyccelMinus(base_shape[i], ind.args[0])
+            else:
+                if isinstance(ind, Slice):
+                    #setting the slice start and end to their correct value if none is provided
+                    start = ind.start
+                    end = ind.end
+                    if ind.start is None:
+                        start = 0
+                    if ind.end is None:
+                        end = expr.base.shape[i]
+                    inds[i] = Slice.__new__(Slice, start, end)
+                #indices of indexedElement of len==1 shouldn't be a Tuple
+                if isinstance(ind, Tuple) and len(ind) == 1:
+                    inds[i].args = ind[0]
+                if allow_negative_indexes and not isinstance(ind, Integer):
+                    inds[i] = PyccelMod(ind, base_shape[i])
+        inds = [self._print(i) for i in inds]
+        #set dtype to the C struct types
+        dtype = self.find_in_dtype_registry(format(expr.dtype), expr.precision, array=True)
+        if expr.rank > 0:
+            return "array_slicing(%s, %s)" % (base, ", ".join(inds))
+        return "%s.nd_%s[get_index(%s, %s)]" % (base, dtype, base, ", ".join(inds))
+
+    def _print_Allocate(self, expr):
+        free_code = ''
+        #free the array if its already allocated
+        if  (expr.status != 'unallocated'):
+            free_code = 'if (%s.raw_data != NULL)\n' % self._print(expr.variable.name)
```

Comment:
This `if` shouldn't be needed if the allocation status is `allocated`
I don't know if it can currently have this status, but it is planned for the future

Reply:
either if its status is "allocated" or "unknown" we need to check if the pointer is not null then free it safely then allocate it without leaking the memory

Reply:
But if the status is `allocated` then the pointer can't be `null`. The `free` is still needed, just not the `if`

Reply:
the `if` is to check if the pointer is **not** `null`

the `_print_Allocate` method is only called if the shapes is different 'that's when the status is unknown'  if the shapes are equal the free is not necessary

Reply:
I agree. But if it **can't** be `null` then the check always returns `True` so you don't need the `if`, just the body of the `if`

Reply:
the "allocated" status will only be needed in the garbage collection step (from what i did understand) but in the case reallocation the status is "unknown" so we need to use `free` because we don't know if its allocated or not

Reply:
Ah not quite. Although currently that will often be true. Let me explain the logic

There are three possible states:
- `unallocated` - This variable has definitely not been allocated it is `null`
   Example 1:
   ```
   nd_array* a;
   // Code which does not use a
   ```
   a is `unallocated`
- `allocated` - This variable has definitely been allocated. It needs to be freed
   Example 2.
   ```
   nd_array* a = new_nd_array(...);
   ```
   a is `allocated`. If we want to allocate it with a different shape it definitely needs freeing. At the end of the function it definitely needs garbage collecting.
- `unknown` - We might have allocated this variable, but I'm not sure
   Example 3.
   ```
   nd_array* a;
   if (b) a = new_nd_array(...);
   ```
   a is `allocated` only if `b`. I do not know `b` so I do not know if a was `allocated`. Therefore the status of a is `unknown`.  If we want to allocate a with a different shape it might need freeing, or it might still be `null`. At the end of the function it might need garbage collecting or it might still be `null`.
   Example 4.
   ```
   nd_array* a;
   for (int i=0; i<n; ++i)
   {
        a = new_nd_array(...);
        // More code here
   }
   ```
   At the end of the for loop a is `allocated` if `n>0`. I don't know `n` therefore the status of a is `unknown`. As above it may or may not need freeing both for re-allocation and for garbage collection
   In addition, when `i=0` a is `unallocated`, but when `i=1` a is `allocated`. Therefore at the allocation line a may or may not be allocated. Therefore the status of a is `unknown` at this line

If you use these definitions your code should be right. The confusion is probably coming from the fact that currently we don't know if we are in an if block or a for loop so when we do:
```
 nd_array* a;
// Mystery code
a = new_nd_array(...);
```
we have no way of knowing what is in the mystery code. We might be in the case of Example 3 or 4. To guard against this we are currently setting the status to `unknown` even when it should be `allocated`. This is something we hope to fix at some point. As far as the printing is concerned though, you don't need to know if the logic was done correctly. You just check the allocation if it is `unknown`. Does that make sense?

Reply:
that's clears a lot of the confusions i had thank you